serve/mlc_serve/engine/staging_engine_worker.py

```diff
@@ -305,20 +316,25 @@ def _adjust_batch(self):
                 # self.max_num_batched_tokens. In such cases, we need to discard the recent decode
                 # tokens that cannot fit into a batch, and recompute them after we fill the cache
                 # entries for the older tokens.
-                if len(self.current_batch) == 0 and num_new_batched_tokens > self.max_num_batched_tokens:
-                    state.token_ids = state.token_ids[:self.max_num_batched_tokens]
-                    state.next_start_position = num_new_batched_tokens = num_tokens = self.max_num_batched_tokens
-                if num_new_batched_tokens > self.max_num_batched_tokens > 0:
+                if (
+                    len(self.current_batch) == 0
+                    and num_new_batched_tokens > self.max_num_batched_tokens
+                ):
+                    state.token_ids = state.token_ids[: self.max_num_batched_tokens]
```

Comment:
We need to define what `max_num_batched_tokens` stands for. Currently it limits the maximum tokens to be processed simultaneously. If we find how to handle this, we can decode until context len. I do not see reasons to cancel, at least just because we achieved `max_num_batched_tokens` limit.

Reply:
> We need to define what max_num_batched_tokens stands for. Currently it limits the maximum tokens to be processed simultaneously.

Yes, exactly. So if the context length (a.k.a., `len(state.token_ids)`) of the request reaches the `max_num_batched_tokens`, in the next engine step, I guess the engine wouldn't select the sequence to put it in the batch since it already reached the batched token limit? So I suspect it will just sit in the queue. 

> But I just wondered about the following case: If there is only sequence in the batch and there is no entries in the cache for other previous sequences, what happens in the generation stage if we max out the free space before that single sequence finishes? We cannot evict since we end up in an infinite loop of evict -> generate -> evict ...

Yes, great point. to prevent this kind of case, seems like we should also consider free space in S2 above.